File: C/board.c
#include <stdio.h>
#include <stdlib.h>

char *init_board(){
    char *game_state = malloc(sizeof(char) * 9);
    for (int i=0; i<9; i++){
        game_state[i] = ' ';
    }
    return game_state;
}

void print_game_state(char game_state[9])
{
    for (int i = 0; i < 9; i++)
    {
        if (i % 3 == 0 && i != 0)
        {
            printf("\n- - -\n");
        }
        printf("%c ", game_state[i]);
    }
    printf("\n");
}

void update_game(char game_state[9], int move_id, char player)
{
    game_state[move_id] = player;
}

int get_total_available_moves(char game_state[9])
{
    int total_empty = 0;
    for (int i = 0; i < 9; i++)
    {
        if (game_state[i] == ' ')
        {
            total_empty++;
        }
    }
    return total_empty;
}

int *get_available_moves(char game_state[9])
{
    // Get available moves from current game state
    int *available_moves = malloc(sizeof(int) * 9);
    int total_available_moves = 0;
    // iterate to find the available move
    for (int i = 0; i < 9; i++)
    {
        if (game_state[i] == ' ')
        {
            available_moves[total_available_moves] = i;
            total_available_moves++;
        }
    }
    return available_moves;
}

void print_available_moves(int available_moves[], int total_available_moves)
{
    printf("Available Moves : ");
    for (int i = 0; i < total_available_moves; i++)
    {
        printf("%d ", available_moves[i]);
    }
    printf("\n");
}

char *copy_board(char game_state[9])
{
    // Copy game board
    char *new_game = malloc(sizeof(char) * 9);
    for (int i = 0; i < 9; i++)
    {
        new_game[i] = game_state[i];
    }
    return new_game;
}

int decide_game(char game_state[], char player)
{
    // 1 : current player win,
    // 0 : Draw,
    // -1 : Continue game
    int win_rules[8][3] = {
        {0, 1, 2},
        {3, 4, 5},
        {6, 7, 8},
        {0, 3, 6},
        {1, 4, 7},
        {2, 5, 8},
        {0, 4, 8},
        {2, 4, 6}};
    int total_count;
    int total_moves = 0;
    int ret = -1;
    // Decide if current player win the game.
    for (int i = 0; i < 8; i++)
    {
        total_count = 0;
        for (int j = 0; j < 3; j++)
        {
            if (game_state[win_rules[i][j]] == player)
            {
                total_count += 1;
            }
        }
        if (total_count == 3)
        {
            ret = 1;
            break;
        }
    }
    // Decide for Draw
    for (int i = 0; i < 9; i++)
    {
        if (game_state[i] != ' ')
        {
            total_moves++;
        }
    }
    if (total_moves == 9 && ret != 1)
    {
        ret = 0;
    }
    // Continue Program
    return ret;
}

File: Python/board.py
class Board:
    def __init__(self):
        pass

    def init_board(self):
        pass

    def updateBoard(self, moveId, player):
        pass

    def decideGame(self, player):
        pass

    def getAvailableMoves(self):
        pass

    def printState(self):
        pass

    def copy(self):
        pass


class TicTacToe(Board):
    def __init__(self):
        self.init_board()
    
    def init_board(self):
        self.game_state =  [' ' for _ in range(9)]

    def updateBoard(self, moveId, player):
        self.game_state[moveId] = player

    def decideGame(self, player):
        win_rules = [[0, 1, 2], [3, 4, 5], [6, 7, 8], [0,3,6], [1, 4, 7], [2, 5, 8], [0,4,8],[2, 4, 6]]
        # 1 for win, 0 for draw, -1 for continue
        ret = -1

        # Decide if current player win condition
        for win_rule in win_rules:
            total_count = 0
            for i in range(3):
                if self.game_state[win_rule[i]] == player:
                    total_count += 1
                    continue
                else:
                    break
            if total_count == 3:
                ret = 1
                break
        
        # Decide for draw condition
        total_moves = 0
        if ret != 1:
            for i in range(9):
                if self.game_state[i] != " ":
                    total_moves += 1
        if total_moves == 9:
            ret = 0
        # Continue the game
        return ret

    def getAvailableMoves(self):
        available_moves = []
        for i in range(9):
            if self.game_state[i] == " ":
                available_moves.append(i)
        return available_moves
    
    def printState(self):
        for i in range(9):
            if i%3 ==0 and i!=0:
                print('\n- - -')
            print(self.game_state[i], end=' ')

    def copy(self):
        game_ = Board()
        for i in range(9):
            if self.game_state[i] != " ":
                game_.updateBoard(i, self.game_state[i])
        return game_
